design/deploy-steps.md

```diff
@@ -0,0 +1,142 @@
+-->
+
+# Customizable deployment procedure
+
+## Status
+
+implementable
+
+## Summary
+
+This design proposes customizing the deployment procedure by providing a high
+level abstraction around Ironic [deploy steps][deploy-steps], as well as
+allowing per-host deployment (IPA) images.
+
+[deploy-steps]: https://docs.openstack.org/ironic/latest/admin/node-deployment.html#deploy-steps
+
+## Motivation
+
+By default the Ironic agent downloads a provided image via HTTP(s), converts it
+(if needed) and writes directly to the disk. This procedure works for many
+distributions, but may not suit some cases well. A prominent example is CoreOS,
+which has its own installer ([coreos-installer][coreos-installer]).
+
+[coreos-installer]: https://github.com/coreos/coreos-installer
+
+### Goals
+
+- A way to replace the default deploy procedure with a custom action.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Add a way to execute custom actions during deployment.
+- A way to provide a link to the Ironic agent image per host.
+
+## Proposal
+
+### User Story
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New structure field `CustomDeploy` with the only field `Method` to engage
+  a non-default deployment procedure as described below.
+- The `Image.URL` field will become optional if `CustomDeploy` is set
+  since we don't know if the deploy method actually requires a separate image.
+
+Setting `CustomDeploy` to a non-empty value will trigger deployment even if
+`Image.URL` is empty.
+
+### Implementation Details: CustomDeploy
+
+When `CustomDeploy` is set, we will use the new [custom-agent deploy
+interface][custom-agent] so that the default deploy steps do not apply. Then we
+request a deploy step with the name provided in the `CustomDeploy.Method`
+field. This step has to be provided in the agent image and is responsible for
+the whole deployment process.
+
+NOTE: `CustomDeploy` is a single-field structure to allow potential extensions
+in the future. The `UserData` and `NetworkData` will be available to a deploy
+step implementation as part of the Node's `instance_info["configdrive"]`.
+
+Any required customization is expected to be provided via `UserData` and
+`NetworkData` or be built into the image itself.
+
+1. A user sets `CustomDeploy.Method = "step_name"` and optionally `Image.URL`,
+   `UserData` and `NetworkData`.
+2. BMO sets `Node.DeployInterface = "custom-deploy"` and requests deployment
+   with a custom deploy step `step_name`.
+3. Ironic boots the agent (if not booted for inspection/cleaning already).
+4. Ironic requests the agent to execute the deploy step `step_name` (it is
+   an error to provide a non-existing step), providing it the complete Node
+   object, including `configdrive`.
+5. Ironic waits for the step to finish.
```

Comment:
What is the expected flow in case of failure? Is there any special way to facilitate the troubleshooting? Also, being a completely custom step I don't think it's possible to perform a preliminary kind of validation for an earlier detection of issues, so I'd expect that most of the problems will be discovered until a later moment (ie a malformed `UserData`)

Reply:
Malformed `UserData` is always discovered very late.

The mode of failure will be pretty much the same as now: the node gets into `deploy failed`. The only new issue is providing a non-existing deploy step, which will also end up in `deploy failed`.